### PR TITLE
fixing ide launch on windows and linux and add tests

### PR DESCRIPTION
**Pull request by fahad-israr**
- Launching ide on windows(incase of Vscode) had an issue on windows as it needed .`cmd` extension to launch.
- Now we have made a separate function that checks if the os is windows.
- Based on that decision we add or don't add `.cmd` (in case of vscode). 
- Added tests for `urlValidator()` and `runCommand()` function .

### CODE REVIEW

File: src/main/java/org/maxandersen/starfish/Starfish.java [hunk 1]
```diff
@@ -72,25 +72,26 @@ public int run(String... args) throws Exception {
...
+    
     String repo_name=args[0].substring(args[0].lastIndexOf("/"),args[0].lastIndexOf(".")); //Extracts the Name of Repository
-    //String clone_path= "/home/fahad/MyProjects/starfish_clonned/"; 
+   
     String originUrl = args[0];
     Path directory = Paths.get(clone_path+repo_name);
+
     if(!Files.exists(directory)) //Check if the user cloned the repo previously and in that case no cloning is needed
-    gitClone(directory, originUrl);
+    gitClone(directory, originUrl);//Calling function to clone the repository
```
Comment (maxandersen) [inline src/main/java/org/maxandersen/starfish/Starfish.java:82]: that comment doesn't say anything beyond what the method name says. 
```suggestion
    gitClone(directory, originUrl);
```

File: src/main/java/org/maxandersen/starfish/Starfish.java [hunk 2]
```diff
@@ -72,25 +72,26 @@ public int run(String... args) throws Exception {
...
    
 
 
 
-    //Launching Vscode on the Cloned Directory 
+    //Launching Editor on the Cloned Directory 
     System.out.println("Launching  Editor Now...");
-    runCommand(directory.getParent(), ide,clone_path+repo_name);
+    launch_editor(directory.getParent(), ide,clone_path+repo_name);//Calling Function to launch Editor
```
Comment (maxandersen) [inline src/main/java/org/maxandersen/starfish/Starfish.java:89]:
```suggestion
    launch_editor(directory.getParent(), ide,clone_path+repo_name);
```

File: src/main/java/org/maxandersen/starfish/Starfish.java [hunk 3]
```diff
@@ -168,6 +169,16 @@ public static void editConfig()throws Exception{
 
 }
 
+//Function to Launch the Editor
+public static void launch_editor(Path directory,String ide,String final_clone_path)throws IOException, InterruptedException{
+//If OS is windows then we add .exe after the command    
+boolean isWindows =System.getProperty("os.name").toLowerCase().indexOf("windows")>=0;
```
Comment (maxandersen) [inline src/main/java/org/maxandersen/starfish/Starfish.java:175]: this is fine for now but this should come from the configuration as a editor command could also be named code.cmd or code.bat on windows thus just adding .exe is not always the right answer..

File: src/main/java/org/maxandersen/starfish/Starfish.java [hunk 4]
```diff
@@ -168,6 +169,16 @@ public static void editConfig()throws Exception{
 
 }
 
+//Function to Launch the Editor
+public static void launch_editor(Path directory,String ide,String final_clone_path)throws IOException, InterruptedException{
```
Comment (maxandersen) [inline src/main/java/org/maxandersen/starfish/Starfish.java:173]: since it is an editor we are assuming to start better to call the parameter `editor` instead of `ide`

File: src/main/java/org/maxandersen/starfish/Starfish.java [hunk 5]
```diff
@@ -168,6 +169,16 @@ public static void editConfig()throws Exception{
...
+boolean isWindows =System.getProperty("os.name").toLowerCase().indexOf("windows")>=0;
+ide=isWindows?ide+".exe":ide;
+runCommand(directory.getParent(), ide,final_clone_path);//Launching the editor now
+
+}
+
+
 public static void gitClone(Path directory, String originUrl) throws IOException, InterruptedException {
     //Function for git clonning
     runCommand(directory.getParent(), "git", "clone", originUrl, directory.getFileName().toString());
```
Comment (maxandersen) [inline src/main/java/org/maxandersen/starfish/Starfish.java:184]: does this work on windows ? shouldn't it be git.exe too ?

Comment (fahad-israr) [reply src/main/java/org/maxandersen/starfish/Starfish.java:184]: This works....It cloned perfectly on my Windows-10 64 bit

**Comment by fahad-israr** [comment]
Fix doesnt work: 
- faced issue:
Screenshot-67
-  Even the CLI doesn't recognize `code.exe` as a command (`'code.exe' is not recognized as an internal or external command,
operable program or batch file.`).

- Instead `code` runs from CLI.
- Also I can see  the** environment variable** `Path` already has the **vscode** configured:
Screenshot (69)

**Comment by fahad-israr** [comment]
- This PR Will not solve the issue
- Issue can be resolved by using cmd.exe
- Another alternative can be by configuring Environment Variables.

Should we close this PR?

File: src/test/java/dk/xam/StarfishTest.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+import static org.hamcrest.CoreMatchers.is;
+
+
+@QuarkusTest
+public class StarfishTest {
+
+    @Test
+    public void testURLValidator() {
+        
+        assertEquals(true,validate_url("https://github.com/user-name/repo-name.git"),"URL Validation test failes");
```
Comment (maxandersen) [inline src/test/java/dk/xam/StarfishTest.java:29]: "failes" is just "fails" - but normally you either leave the description empty or describe what was validated. i.e. `assertEquals(true, validate_url("https://github.com/user-name/repo-name.git"), "Plain github url"))`

File: src/test/java/dk/xam/StarfishTest.java [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+    @Test
+    public void testLaunchEditor()throws Exception {
+    //To test launch_editor(dir,ide,dir) function
+        String userHome = System.getProperty( "user.home" ); 
+        String expected = "";
+        Path directory = Paths.get(userHome);
+
+       
+        if(isWindows()){
+            assertEquals(expected,launch_editor(directory,"code.cmd",userHome),"Editor Launch Test Failed");
```
Comment (maxandersen) [inline src/test/java/dk/xam/StarfishTest.java:49]: this assumes code.cmd is installed on the machine running tests. Pretty sure github action will not have code.cmd on its path.

File: src/test/java/dk/xam/StarfishTest.java [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+            
+      
+           
+        }
+        else
+        {
+           assertEquals(expected,launch_editor(directory,"code",userHome),"Editor Launch Test Failed");
```
Comment (maxandersen) [inline src/test/java/dk/xam/StarfishTest.java:56]: same - code is not guaranteed to exists. you'll need to find another way of testing launch than using code.exe/code. maybe ensure the ci setup has a command you can call?

File: src/test/java/dk/xam/StarfishTest.java (hunk 6 above)
Comment (fahad-israr) [reply src/test/java/dk/xam/StarfishTest.java:29]: Ugh a typo 
"Plain github url" : yup!! I'll put in that

File: src/test/java/dk/xam/StarfishTest.java (hunk 8 above)
Comment (fahad-israr) [reply src/test/java/dk/xam/StarfishTest.java:56]: " ensure the ci setup has a command you can call?" - Looking for this now.

File: src/main/java/dk/xam/starfish/Starfish.java [hunk 9]
```diff
@@ -187,22 +213,30 @@ public static void runCommand(Path directory, String... command) throws IOExcept
...
+    int exit = p.waitFor();
+
+    if (exit != 0) {
+        throw new AssertionError(String.format("runCommand returned %d", exit));
+    }
+    return p;
+}
+
+
+public static String gobbleStream(Process p)throws IOException, InterruptedException{
```
Comment (maxandersen) [inline src/main/java/dk/xam/starfish/Starfish.java:226]:
```suggestion
public static String gobbleStream(Process p) throws IOException, InterruptedException{
```

**Comment by fahad-israr** [comment]
- Removed echo function 
- Added test for runCommand() function

File: installer/linux/installer/Readme.md [hunk 10]
```diff
@@ -0,0 +1,2 @@
+# Protocol Handler for `ide://...` scheme.
```
Comment (maxandersen) [inline installer/linux/installer/Readme.md:1]: Lets be consistent - don't use mixed casing in files and use the same format. we have `README.adoc` in root, lets use same here.

File: installer/windows/installer/install.cmd [hunk 11]
```diff
@@ -0,0 +1,7 @@
+@echo off
```
Comment (maxandersen) [inline installer/windows/installer/install.cmd:1]: this is a comment to the .keep file - why is that file here ?

Comment (maxandersen) [reply installer/windows/installer/install.cmd:1]: ah its for osx. never mind then.

Comment (fahad-israr) [reply installer/windows/installer/install.cmd:1]: Like I was able to develop Protocol Handler for Linux and Windows I have to do for mac as well.
to maintain directory structure on github I have that keep file.

File: src/main/java/dk/xam/starfish/Starfish.java [hunk 12]
```diff
@@ -187,22 +213,30 @@ public static void runCommand(Path directory, String... command) throws IOExcept
...
     StreamGobbler errorGobbler = new StreamGobbler(p.getErrorStream(), "E");
     StreamGobbler outputGobbler = new StreamGobbler(p.getInputStream(), "O");
     outputGobbler.start();
     errorGobbler.start();
-    int exit = p.waitFor();
```
[the remark this answers is not in the document]
Comment (maxandersen) [reply src/main/java/dk/xam/starfish/Starfish.java:194]: this is now fixedand we can resolve it?
